big_tests/tests/mam_proper_SUITE.erl

```diff
@@ -0,0 +1,151 @@
+
+all() ->
+    [{group, async_writer}].
+
+groups() ->
+    [{async_writer, [], [async_writer_store]}].
+
+init_per_suite(C) ->
+    application:ensure_all_started(jid),
+    escalus:init_per_suite(C).
+
+end_per_suite(C) ->
+    escalus:end_per_suite(C).
+
+init_per_group(_G, C) ->
+    case mongoose_helper:is_rdbms_enabled(domain_helper:host_type()) of
+        true -> C;
+        false -> {skip, "rdbms not enabled"}
+    end.
+
+end_per_group(_G, C) ->
+    C.
+
+init_per_testcase(Name, C) ->
+    escalus:init_per_testcase(Name, C).
+
+end_per_testcase(Name, C) ->
+    escalus:end_per_testcase(Name, C).
+
+usernames() ->
+   [<<"alice">>, <<"bob">>, <<"bob.1">>, <<"bob44">>, <<"a">>, <<"xd">>,
+    <<"admin">>].
+
+resources() ->
+    [<<"res1">>, <<"PICKLE">>, <<"1">>, <<"xdxd">>].
+
+origin_id() ->
+    oneof([none, <<"orig_id">>, proper_types:non_empty(proper_unicode:utf8(5))]).
+
+username() ->
+    ?SUCHTHAT(U, proper_types:non_empty(oneof(usernames())),
+              jid:nodeprep(U) =/= error).
+
+hostname() ->
+    proper_types:oneof([<<"localhost">>, <<"otherhost">>, <<"example.com">>]).
+
+resource() ->
+    ?SUCHTHAT(R, proper_types:non_empty(oneof(resources())),
+              jid:resourceprep(R) =/= error).
+
+jid() ->
+    ?SUCHTHAT(Jid, maybe_invalid_jid(), Jid =/= error).
+
+maybe_invalid_jid() ->
+    ?LET({U, S, R},
+         {username(), hostname(), resource()},
+         make_jid(U, S, R)).
+
+make_jid(U, S, R) when is_binary(U), is_binary(S), is_binary(R) ->
+    mongoose_helper:make_jid(U, S, R).
+
+direction() ->
+    proper_types:oneof([incoming, outgoing]).
+
+body() ->
+    proper_unicode:utf8(1000).
+
+packet(To, Body) ->
+    escalus_stanza:chat_to(jid:to_binary(To), Body).
+
+%% Generates mod_mam:archive_message_params()
+params() ->
+    ?LET({MessId, ArcId, LocalJid, RemoteJid,
+          OriginId, Dir, Body, SenderId},
+         {non_neg_integer(), non_neg_integer(), jid(), jid(),
+          origin_id(), direction(), body(), non_neg_integer()},
+         #{message_id => MessId, archive_id => ArcId,
+           local_jid => LocalJid, remote_jid => RemoteJid,
+           source_jid => choose_source_jid(Dir, LocalJid, RemoteJid),
+           origin_id => OriginId, direction => Dir,
+           packet => packet(choose_dest_jid(Dir, LocalJid, RemoteJid), Body),
+           sender_id => SenderId}).
+
+choose_source_jid(incoming, _LocalJid, RemoteJid) -> RemoteJid;
+choose_source_jid(outgoing, LocalJid, _RemoteJid) -> LocalJid.
+
+choose_dest_jid(outgoing, _LocalJid, RemoteJid) -> RemoteJid;
+choose_dest_jid(incoming, LocalJid, _RemoteJid) -> LocalJid.
+
+params_list() ->
+    ?LET(ParamsList, proper_types:non_empty(proper_types:list(params())),
+         with_unique_message_id(ParamsList)).
+
+with_unique_message_id(ParamsList) ->
+    Pairs = [{MessId, Params} || #{message_id := MessId} = Params <- ParamsList],
+    %% Removes duplicates
+    maps:values(maps:from_list(Pairs)).
+
+async_writer_store(C) ->
```

Comment:
Please reorder functions so that the test case is not in the middle of helpers. It is difficult to figure out where the test case is.